packages/schema/src/query_response.rs

```diff
@@ -0,0 +1,152 @@
+use std::collections::{BTreeMap, BTreeSet};
+
+use schemars::{schema::RootSchema, JsonSchema};
+use thiserror::Error;
+
+pub use cosmwasm_schema_derive::QueryResponses;
+
+/// A trait for tying QueryMsg variants (different contract queries) to their response types.
+/// This is mostly useful for the generated contracted API description when using `cargo schema`.
+///
+/// Using the derive macro is the preferred way of implementing this trait.
+///
+/// # Example
+/// ```
+/// use cosmwasm_schema::QueryResponses;
+/// use schemars::JsonSchema;
+///
+/// #[derive(JsonSchema)]
+/// struct AccountInfo {
+///     IcqHandle: String,
+/// }
+///
+/// #[derive(JsonSchema, QueryResponses)]
+/// enum QueryMsg {
+///     #[returns(Vec<String>)]
+///     Denoms {},
+///     #[returns(AccountInfo)]
+///     AccountInfo { account: String },
+/// }
+/// ```
+pub trait QueryResponses: JsonSchema {
+    fn response_schemas() -> Result<BTreeMap<String, RootSchema>, IntegrityError> {
+        let response_schemas = Self::response_schemas_impl();
+
+        let queries: BTreeSet<_> = response_schemas.keys().cloned().collect();
+
+        check_api_integrity::<Self>(queries)?;
+
+        Ok(response_schemas)
+    }
+
+    fn response_schemas_impl() -> BTreeMap<String, RootSchema>;
+}
+
+/// `generated_queries` is expected to be a sorted slice here!
+fn check_api_integrity<T: QueryResponses + ?Sized>(
+    generated_queries: BTreeSet<String>,
+) -> Result<(), IntegrityError> {
+    let schema = crate::schema_for!(T);
+
+    // something more readable below?
+
+    let schema_queries: BTreeSet<_> = schema
+        .schema
+        .subschemas
+        .ok_or(IntegrityError::InvalidQueryMsgSchema)?
+        .one_of
+        .ok_or(IntegrityError::InvalidQueryMsgSchema)?
+        .into_iter()
+        .map(|s| {
+            s.into_object()
+                .object
+                .ok_or(IntegrityError::InvalidQueryMsgSchema)?
+                .required
+                .into_iter()
+                .next()
+                .ok_or(IntegrityError::InvalidQueryMsgSchema)
+        })
+        .collect::<Result<_, _>>()?;
+    if schema_queries != generated_queries {
+        return Err(IntegrityError::InconsistentQueries {
+            query_msg: schema_queries,
+            responses: generated_queries,
+        });
+    }
+
+    Ok(())
+}
+
+#[derive(Debug, Error, PartialEq)]
+pub enum IntegrityError {
+    #[error("the structure of the QueryMsg schema was unexpected")]
+    InvalidQueryMsgSchema,
+    #[error(
+        "inconsistent queries - QueryMsg schema has {query_msg:?}, but query responses have {responses:?}"
+    )]
+    InconsistentQueries {
+        query_msg: BTreeSet<String>,
+        responses: BTreeSet<String>,
+    },
+}
+
+#[cfg(test)]
+mod tests {
+    use schemars::schema_for;
+
+    use super::*;
+
+    #[derive(Debug, JsonSchema)]
+    #[serde(rename_all = "snake_case")]
+    #[allow(dead_code)]
```

Comment:
Is this still needed when `GoodMsg` is in the tests block?

Reply:
Yes, the compiler complaints about the enum variants not being used. Makes sense since they're only interpreted by macros.